### ISSUE TITLE
Develop a New DbGaP Data Dictionary Parser

### ISSUE THREAD

**Issue by stevencox** [helxplatform/dbgaptools#1]
Develop a new version of the [dbGaP data dictionary parser](https://github.com/heliumplusdatastage/dbgaptools/blob/master/R/read_functions.R)

The new implementation should:
* **Python**: Use Python 3.7.x
* **Unit Test**: Develop unit tests following the tests in the existing code as closely as possible to provide an ability to verify the correctness of the new implementation.
* **Travis**: Execute unit tests via Travis.
* **Generality**: The current version has limitations we'll further characterize and resolve as this work progresses. Adrienne Stilp, the code's author suggests we rewrite it. From Adrienne: ```"it is geared at investigators preparing files for dbGaP. That leads to some oddities in the format of the returned data frame (in particular, storing N encoded value definitions in the N last columns). It also doesn't store everything you might want such as phv identifiers, but it should help get you started."```
* **Containerize**: Create a Docker image containerizing the parser.
* **Security Scan**: Prefer a base container with a well understood path to security remediation. The developed container needs to pass the security vetting machinery, currently Anchore.

@diatomsRcool

**Comment by stevencox** [comment]
See this [repo](https://github.com/NCATS-Tangerine/tranql/tree/master/tranql) as an example of using pytest and Travis.

**Comment by stevencox** [comment]
A Python parser shell to get you started:
```
import json
import yaml
import os
import xml.etree.ElementTree as ET

tree = ET.parse('data.xml')
root = tree.getroot()

output = {}
for data in root.iter('data'):
    for c in data.iter ('country'):
        #print(c.attrib['name'])
        node = {}
        output['country'] = node
        node['name'] = c.attrib['name']
        for e in c.iter (): 
            node[e.tag] = e.text

print (f"{json.dumps(output, indent=2)}")

```

**Comment by diatomsRcool** [comment]
Partially blocked by #3

**Comment by muralikarthikk** [comment]
Data Dictionary Parser using python.

read_dd_xml function converts the XML document to a nice pandas data frame.

Sample file used:
This file can be downloaded from the dbgap ftp,
```ftp://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs000988/phs000988.v2.p1/pheno_variable_summaries/phs000988.v2.pht005248.v2.TOPMed_WGS_Asthma_Costa_Rica_Subject_Phenotypes.data_dict.xml```

```
import pandas as pd
import xml.etree.ElementTree as ET

def read_dd_xml(filename):
    # Set parent_dd_file to the filename of the XML data dictionary on disk
    xml_dd = ET.parse(filename)

    # Select variable nodes
    variable_nodes = xml_dd.findall('variable')

    # Create a one-line data frame for each variable node
    required_nodes = {
        'VARNAME': 'name',
        'VARDESC': 'description'
    }

    # Process some optional nodes; others are ignored
    optional_nodes = {
        'TYPE': 'type',
        'UNITS': 'unit',
        'MIN': 'logical_min',
        'MAX': 'logical_max'
    }

    # unique_keys = xml_dd.findall('unique_key')
    # print(unique_keys)

    df_list = []
    for variable_node in variable_nodes:
        df = {}

        # Search for required nodes on each variable node and add them to the df dictionary.
        for key, value in required_nodes.items():
            xpath = './/%s' % value
            text = variable_node.find(xpath).text
            df[key] = text

        # Search for optional nodes on each variable node and add them to the df dictionary.
        for key, value in optional_nodes.items():
            xpath = './/%s' % value
            node = variable_node.find(xpath)
            if node is not None:
                df[key] = node.text

        # if len(unique_keys) > 0:
        #     for value in unique_keys:
        #         if df['VARNAME'] == value.text:
        #             df['UNIQUEKEY'] = 'X'

        # Find all the value nodes in each variable node. If it has multiple value nodes. Value with index 0 - VALUES
        # Value with index 1 - X__1
        child_value_nodes = variable_node.findall('.//value')
        for index, value in enumerate(child_value_nodes):
            try:
                value_string = str(value.attrib['code']) + '=' + value.text
                if index == 0:
                    name_string = 'VALUES'
                else:
                    name_string = 'X__' + str(index)
                df[name_string] = value_string
            except Exception as _:
                pass

        # Append the df dictionary of each variable node to the df list.
        df_list.append(df)

    # Create a dataframe from the df list.
    dd = pd.DataFrame(df_list)

    required_column_order = ['VARNAME', 'VARDESC', 'TYPE', 'UNITS', 'MIN', 'MAX', 'UNIQUEKEY', 'VALUES']

    # Arrange the dataframe columns in a specific order.
    ordered_dd = pd.DataFrame()
    for value in required_column_order:
        if value in dd.columns:
            ordered_dd[value] = dd[value]
            dd.pop(value)

    # Concatenates the left over columns on dd dataframe to the ordered_dd dataframe. Return the ordered dataframe.
    ordered_dd = pd.concat([ordered_dd, dd], axis=1, sort=False)
    return ordered_dd


result = read_dd_xml('/path/to/the/sample/xml/file')
pd.set_option('expand_frame_repr', False)
pd.set_option('display.max_columns', 999)
print(result)
```

Output:
<img width="700" alt="xml_dataframe" src="https://user-images.githubusercontent.com/41588376/55355469-5f2b1f80-5496-11e9-9a49-9dcb19afbd60.PNG">

**Comment by stevencox** [comment]
Hi @muralikarthikk - this looks good - please create a folder for it in the repo and check it in.

**Comment by muralikarthikk** [comment]
Hi @stevencox - Created a new branch "development" and created a folder "python" in the repo.

**Comment by stevencox** [comment]
For the development team:

```
> Hi Steve,
> 
> I  prefer email for larger questions since it's easier to find, save, and forward information, especially since it looks like the current Data STAGE Slack doesn't have the full message history. That said, I am a member of the nhlbi-datastage slack, so it's fine for quick questions.
> 
> The data dictionary parsing is very straightforward. We've written some tools to parse them in R, which could be a good reference for re-implementing a parser that returns your preferred format. In particular, they use Xpath to find specific types of child nodes (variable names, descriptions, identifiers, etc). You'll want to look at the .read_dd_xml function in the dbgaptools package on github:
> https://github.com/UW-GAC/dbgaptools/blob/master/R/read_functions.R
> 
> I wouldn't recommend using this function directly in Data STAGE, since it is geared at investigators preparing files for dbGaP. That leads to some oddities in the format of the returned data frame (in particular, storing N encoded value definitions in the N last columns). It also doesn't store everything you might want such as phv identifiers, but it should help get you started.
> 
> There is also an appendix in the white paper we put together for the NIH Data Commons that shows some more examples of data dictionaries in R:
> https://docs.google.com/document/d/1e77a7dZ9I1FcXHbK-CXkbXmP5xywsw1dePsoxX_E6OM/edit#heading=h.j4r4bxlk382q
> 
> And finally, less relevant immediately, but there's a new section in the White Paper about issues we encountered trying to read files (mostly data files) from dbGaP:
> https://docs.google.com/document/d/1e77a7dZ9I1FcXHbK-CXkbXmP5xywsw1dePsoxX_E6OM/edit#heading=h.1oymh7ljual0
> 
> 
> Thanks,
> Adrienne
```

**Comment by bheavner** [comment]
This tool should be something that can run on dataSTAGE platforms (e.g. terra, seven bridges). I'm not sure how we make that happen, but suggest that it be something that can be dockerized and fit into a workflow as a target deployment approach. I can imagine a container expecting a data dictionary file (or path?) as an input, producing the index 1 file as an output. I think having it in WDL covers terra. I'm not sure how we do it in seven bridges.

**Comment by diatomsRcool** [comment]
This tool has been user and performance tested. I'm closing this issue. If any problems start up, they can be listed in a new issue.
